### PR TITLE
Reverse index creation and retrieval

### PR DESCRIPTION
**Pull request by timwangmusic**
## Description
Kick start the effort to improve solution generation speed using Reverse Index.

## Solution
* The first commit in this PR shortens RedisClient receiver names per Go readability recommendations.
* Define reverse index and its metadata.
* Define methods to save/retrieve indexes with Redis sorted set.

## Testing
- [ ] Integration testing on Heroku staging
- [x] Added new unit tests

## Checks
- [x] Have you removed commented code?
- [x] Have you used gofmt to format your code?

### CODE REVIEW

File: iowrappers/reverse_index.go [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *RedisClient) RetrieveReverseIndex(ctx context.Context, metadata *IndexMetadata, count int64) (*ReverseIndex, error) {
+	redisKey := reverseIndexKey(metadata)
+	// only limit the number of places
+	res := r.Get().ZRevRangeByScore(ctx, redisKey, &redis.ZRangeBy{
```
Comment (rwangsc18) [inline iowrappers/reverse_index.go:47]: Is the retrieval done with sorted scores?

**Review by rwangsc18 — APPROVED** [review]
LGTM. left one question.

**Comment by rwangsc18** [comment]
> LGTM. left one question.

Very nice work. I will spend some time catching up with sorted set reference that you shared about Redis.

File: iowrappers/reverse_index.go (hunk 1 above)
Comment (timwangmusic) [reply iowrappers/reverse_index.go:47]: Yes, the retrieved places are sorted with score from high to low.

**Pull request merged**